### DISCUSSION TITLE
[SOLVED] Using dynamically compiled packages without installing full Entware

### DISCUSSION THREAD

**Discussion by puterboy** [Entware/Entware#821] [Q&A] [answered]
I would like to be able to run small simple executables compiled under Entware on devices that do not have a full Entware support.

Since it seems like Entware gcc doesn't support a static libc, I was hoping I could just copy over the required dyamic libraries from an identical source device with the full Entware install and then set LD_LIBRARY_PATH if necesary to dynamically link and run my executable.
Even better since LD_LIBRARY_PATH on the target machine already includes /jffs/lib which is writable, I just copied the dynamic libraries there.

```
$ echo $LD_LIBRARY_PATH
/lib:/usr/lib:/jffs/lib:/jffs/usr/lib:/jffs/usr/local/lib:/mmc/lib:/mmc/usr/lib:/opt/lib:/opt/usr/lib
```

I used 'ldd' to determine the appropriate libraries to copy over.
```
$ ldd ./myexecutable
 linux-vdso.so.1 (0x7ee8d000)
 libgcc_s.so.1 => /opt/lib/libgcc_s.so.1 (0x76eea000)
 libc.so.6 => /opt/lib/libc.so.6 (0x76da8000)
 /opt/lib/ld-linux.so.3 (0x76f07000)
```
I copied over libgcc_s.so.1, libc.so.6 and ld-linux.so.3 (linux-vdso.so.1 doesn't exist on the source machine and the executable still works there)

```
$ ls -al /jffs/lib
total 1400
drwxr-xr-x    2 root     root             0 Jul 17 23:55 .
drwxr-xr-x    6 root     root             0 Jul 17 22:30 ..
-rwxr-xr-x    1 root     root        134428 Jul 17 22:39 ld-2.27.so
lrwxrwxrwx    1 root     root            10 Jul 17 23:55 ld-linux.so.3 -> ld-2.27.so
-rwxr-xr-x    1 root     root       1247408 Jul 17 22:39 libc-2.27.so
lrwxrwxrwx    1 root     root            12 Jul 17 23:55 libc.so.6 -> libc-2.27.so
-rw-r--r--    1 root     root         50848 Jul 17 22:39 libgcc_s.so.1
```

Unfortunately, whatever I try on the target machine, I still get:
```
$ ./myexecutable
-sh: ./myexecutable: not found
$ ls -al ./myexecutable`
-rwxr-xr-x    1 root     root         13576 Jul 17 22:42 ./myexecutable
```
What am I missing????
Is this possible to do?
I am not able to install entware itself on the other machines because I lack sufficient space.

Thanks!

**Comment by zyxmon** [comment, marked as answer]
Entware binaries need system libraries and dynamic linker (ld*) to be in /opt/lib folder

**Comment by puterboy** [comment]
So, is there any way to create executables that don't require an /opt/lib install?

**Comment by zyxmon** [comment]
There is no way. You can try chroot environment - but it will require some extra space - https://forums.zyxmon.org/viewtopic.php?f=5&t=5797 (Russian)

**Reply by puterboy** [reply]
> There is no way. You can try chroot environment - but it will require some extra space - https://forums.zyxmon.org/viewtopic.php?f=5&t=5797 (Russian)

Thanks! Very helpful
Actually, did not need a chroot environment.

I simply bind-mounted /jffs/lib on /opt/lib (when not using entware this directory has no libs in it)
```
mount --rbind /jffs/lib /opt/lib
```
and then put the needed libraries in /jffs/lib which is persistent... very simple!

To automate, the bind-mount can either
1. Be added to the startup (potentially checking first to make sure that entware not running and/or that no libs in stored in /opt/lib)
2. Be added to a wrapper script that first does the bind mount before running the routine and then unmounts after

Really simple solution and requires no more extra space than the space to store the dynamic libraries referenced by the compiled routines.

**Comment by puterboy** [comment]
Alternatively, are there any simple ways of creating statically-linked binaries?
It seems like the standard gcc with Entware lacks a static libc library for linking

I would think this is a common use case where people may want to create a simple standalone executable but don't need the full power of an Entware install. 

Surely, I'm not the first person to have this need :)

**Comment by zyxmon** [comment]
https://github.com/Entware/Entware/wiki/How-to-add-a-new-package
opkg is an example of statically built binary
